### PR TITLE
Adding csv output

### PR DESCRIPTION
**Pull request by psteinb**
adds CLI flag for producing csv output of the measurement summary. the code becomes a bit more complicated with this, but the csv provides more added value downstream if the results are to be plotted

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -42,6 +42,8 @@ unsigned int ARRAY_SIZE = 33554432;
 unsigned int num_times = 100;
 unsigned int deviceIndex = 0;
 bool use_float = false;
+bool output_as_csv = false;
+std::string csv_seperator = ",";
```
Comment (Ruyk) [inline main.cpp]: shouldn't this be csv_separator instead?

**Comment by Ruyk** [comment]
Beware! that the typo is in other places as well, maybe do a find/replace?

**Comment by tomdeakin** [comment]
In general this looks fine, but there are some formatting issues with your conditional statements throughout. For example:

```if(!output_as_csv){```

This should be like the following, with a space between `if` and `(` and the `{` on it's own line aligned with the `if`.
```
if (!output_as_csv)
{
```

**Comment by psteinb** [comment]
just pushed an update. feel free to have a look.

**Comment by tomdeakin** [comment]
There are still a few mistakes, so I'll go through it line by line to point them out.

File: main.cpp [hunk 2]
```diff
@@ -179,14 +192,30 @@ void run()
...
+      << "function"  << csv_separator
+      << "nreps"  << csv_separator
+      << "n_elements"  << csv_separator
+      << "sizeof"  << csv_separator
+      << "max_mbytes_per_sec"<< csv_separator
+      << "min_runtime" << csv_separator
+      << "max_runtime" << csv_separator
+      << "avg_runtime" << std::endl;
+  }
+  else{
```
Comment (tomdeakin) [inline main.cpp]: Fix style

File: main.cpp [hunk 3]
```diff
@@ -206,14 +235,28 @@ void run()
...
+        << num_times << csv_separator
+        << ARRAY_SIZE << csv_separator
+        << sizeof(T) << csv_separator
+        << 1.0E-6 * sizes[i] / (*minmax.first) << csv_separator
+        << *minmax.first << csv_separator
+        << *minmax.second << csv_separator
+        << average << csv_separator
+        << std::endl;
+    }
+    else {
```
Comment (tomdeakin) [inline main.cpp]: Fix style

File: main.cpp [hunk 4]
```diff
@@ -179,14 +192,30 @@ void run()
   check_solution<T>(num_times, a, b, c, sum);
 
   // Display timing results
-  std::cout
-    << std::left << std::setw(12) << "Function"
-    << std::left << std::setw(12) << "MBytes/sec"
-    << std::left << std::setw(12) << "Min (sec)"
-    << std::left << std::setw(12) << "Max"
-    << std::left << std::setw(12) << "Average" << std::endl;
+  if(output_as_csv)
```
Comment (tomdeakin) [inline main.cpp]: No space after `if`

File: main.cpp [hunk 5]
```diff
@@ -206,14 +235,28 @@ void run()
     double average = std::accumulate(timings[i].begin()+1, timings[i].end(), 0.0) / (double)(num_times - 1);
 
     // Display results
-    std::cout
+    if(output_as_csv)
```
Comment (tomdeakin) [inline main.cpp]: No space after `if`

File: main.cpp [hunk 6]
```diff
@@ -206,14 +235,28 @@ void run()
...
+        << std::endl;
+    }
+    else {
+      std::cout
       << std::left << std::setw(12) << labels[i]
       << std::left << std::setw(12) << std::setprecision(3) << 1.0E-6 * sizes[i] / (*minmax.first)
       << std::left << std::setw(12) << std::setprecision(5) << *minmax.first
       << std::left << std::setw(12) << std::setprecision(5) << *minmax.second
       << std::left << std::setw(12) << std::setprecision(5) << average
       << std::endl;
```
Comment (tomdeakin) [inline main.cpp]: Lines 253 to 258 should be indented

**Comment by psteinb** [comment]
another updated based on your comments. is there a linter config that you can share so that stuff like this could be done automatically?

File: main.cpp [hunk 7]
```diff
@@ -179,14 +192,31 @@ void run()
...
-    << std::left << std::setw(12) << "Function"
-    << std::left << std::setw(12) << "MBytes/sec"
-    << std::left << std::setw(12) << "Min (sec)"
-    << std::left << std::setw(12) << "Max"
-    << std::left << std::setw(12) << "Average" << std::endl;
+  if (output_as_csv)
+  {
+    std::cout
+      << "function"  << csv_separator
+      << "nreps"  << csv_separator
```
Comment (tomdeakin) [inline main.cpp]: Can you change this to be `num_times` please? It's more consistent with the CLI arguments and the style of using underscores to separate words in the header.

File: main.cpp [hunk 8]
```diff
@@ -196,7 +196,7 @@ void run()
   {
     std::cout
       << "function"  << csv_separator
-      << "nreps"  << csv_separator
+      << "num_times"  << csv_separator
       << "n_elements"  << csv_separator
       << "sizeof"  << csv_separator
       << "max_mbytes_per_sec"<< csv_separator
```
Comment (tomdeakin) [inline main.cpp]: Missing space.

File: main.cpp [hunk 9]
```diff
@@ -199,7 +199,7 @@ void run()
       << "num_times"  << csv_separator
```
Comment (tomdeakin) [inline main.cpp]: Lines 199-201 have 2 spaces between the string and the `<<`.

File: main.cpp [hunk 10]
```diff
@@ -179,14 +192,31 @@ void run()
...
-  std::cout
-    << std::left << std::setw(12) << "Function"
-    << std::left << std::setw(12) << "MBytes/sec"
-    << std::left << std::setw(12) << "Min (sec)"
-    << std::left << std::setw(12) << "Max"
-    << std::left << std::setw(12) << "Average" << std::endl;
+  if (output_as_csv)
+  {
+    std::cout
+      << "function"  << csv_separator
```
Comment (tomdeakin) [inline main.cpp]: There is still extra spaces on line 198 and 199.

**Comment by tomdeakin** [comment]
Looks great now, thanks for your patience! I'll merge now.

**Pull request merged**